lib/cdo/optimizer.rb

```diff
@@ -0,0 +1,103 @@
+require 'timeout'
+require 'image_optim'
+require 'image_compressor_pack'
+require 'digest/md5'
+require 'active_job'
+require 'active_support/core_ext/module/attribute_accessors'
+require 'active_support/cache'
+
+# Optimizes content on-the-fly based on provided content-type.
+# If the process takes longer than specified timeout, the original data will be returned
+# and the optimization will finish asynchronously.
+module Cdo
+  class Optimizer
+    MIME_TYPES = %w[
+      image/gif
+      image/jpeg
+      image/png
+      image/svg+xml
+    ].freeze
+
+    # Since optimization steps can take a long time,
+    # cache results in persistent storage to avoid duplicate work.
+    # Default to Rails.cache if it's a File or MemCache store.
+    # Otherwise, create a new FileStore for local use.
+    mattr_accessor :cache do
+      rails_cache = (defined?(Rails) && Rails.cache)
+      if rails_cache.is_a?(ActiveSupport::Cache::FileStore) ||
+        rails_cache.is_a?(ActiveSupport::Cache::MemCacheStore)
+        rails_cache
+      else
+        ActiveSupport::Cache::FileStore.new(dashboard_dir('tmp', 'cache'))
+      end
+    end
+
+    # Set timeout to non-zero to wait the specified number of seconds for the
+    # optimization to finish, before returning nil.
+    # Default no timeout.
+    mattr_accessor(:timeout) {0}
+
+    # @param data [String] input content
+    # @param content_type [String] content type
+    # @return [String] optimized content (or nil if optimization is pending)
+    def self.optimize(data, content_type)
+      Timeout.timeout(timeout) do
+        case content_type
+          when *MIME_TYPES
+            optimize_image(data)
+          else
+            raise 'Invalid content type'
+        end
+      end
+    rescue Timeout::Error
+      # Optimization is still pending after timeout.
+      nil
+    end
+
+    SLEEP_INTERVAL = 0.1
```

Comment:
I was surprised a RuboCop rule didn't enforce this style-guide rule (ref [`consistent-classes`](https://github.com/bbatsov/ruby-style-guide#consistent-classes)), it seems a Cop was never implemented for it. I'll leave as-is for now since following standard class-structure loses a slight bit of context in this case, and we're not (yet) rigidly enforcing the `consistent-classes` rule through automation.